### ISSUE TITLE
Julia language as a native citizen?

### ISSUE THREAD

**Issue by e-kotov** [posit-dev/positron#3679]
Since it is a data science IDE now, not just R, or Python. Any plans for Julia support?
Currently it is awkward that even though VSCode internals of Positron recognise Julia code and highlight it, but the choice of kernels is limited to R and Python (I have Julia installed).

<img width="903" alt="Screenshot 2024-06-28 at 08 10 54" src="https://github.com/posit-dev/positron/assets/8681379/55871cf1-ce74-4f30-b0ba-81aa4c1c564e">

**Comment by pat-alt** [comment]
Just learned about this project and this was the first question that came to my mind. I'm generally quite happy with the Julia VSCode extension, but would love to test Positron for Julia.

**Comment by juliasilge** [comment]
Thank you so much for your interest in Positron! 🙌 Positron is in fact built with the kind of extensibility that would allow us to add support for Julia, or other data science languages that become important in the future. Positron's API for extensions is not fully mature yet so supporting another language would be a bit further in the future than we are thinking right now.

**Comment by chlazaris** [comment]
It would be great indeed to offer native support for Julia and any other language important to Data Science. Thank you everyone and @juliasilge. Looking forward!

**Comment by wesm** [comment]
If anyone in the community is interested in looking at this -- aside from needing a Jupyter kernel (which Julia has: https://github.com/JuliaLang/IJulia.jl), a place to start would be to implement a code generator to create Julia RPC handlers for the custom messages described here

https://github.com/posit-dev/positron/tree/main/positron/comms

Code that uses these custom Jupyter comm handlers can be found in the positron_ipykernel for Python or the Ark R kernel

**Comment by seeM** [comment]
In addition to Wes's advice, it should also be possible to hook up _any_ Jupyter kernel to Positron by creating an extension that calls `positron.runtime.registerLanguageRuntimeManager` and implements `positron.LanguageRuntimeManager` and `positron.LanguageRuntimeSession` as needed. The R extension would be a good reference: https://github.com/posit-dev/positron/tree/main/extensions/positron-r.

We would like to make the above easier for arbitrary Jupyter kernels in future (https://github.com/posit-dev/positron/issues/3585).

You should be able to do that before even implementing the additional comms Wes mentioned, and you should then have basic access to at least the console pane. From there, you can implement the comms to get additional support for the rest (variables, plots, help, data explorer, etc).

Do note, however, that third-party language packs are still experimental:

1. We have not yet tested this approach with external extensions (i.e. not bundled with Positron).
2. We are likely to make breaking changes to the `positron` extension API at this stage.

If you try this and get stuck, please feel free to post here; I'm happy to give some guidance!

**Comment by nicomakesdirt** [comment]
Seconding interest in Julia!

**Comment by liu20001028** [comment]
please add julia support

**Comment by ntluong95** [comment]
Please bring Julia to the game

**Comment by liu20001028** [comment]
Did you planned add julia

**Comment by multidis** [comment]
+1 for Julia support

**Comment by AcademicMJS** [comment]
Is there any official or unofficial information regarding the ongoing or planned support for Julia in Positron?

Are there any known workarounds or beta features available for Julia users who want to integrate it smoothly with Positron? 😄

**Comment by juliasilge** [comment]
To give a little update since July, the situation is still the same as outlined by:

- Wasim in https://github.com/posit-dev/positron/issues/3679#issuecomment-2202599985 on the need for an extension to implement the Positron language runtime infrastructure
- Wes in https://github.com/posit-dev/positron/issues/3679#issuecomment-2201065164 on the need for RPC handlers

To reiterate what Wasim shared, we are definitely up for collaboration and help on the needed steps! It is still important to keep in mind that Positron's extension API is experimental at the current time.

**Comment by trafficfan** [comment]
add 1 for Julia support

**Comment by Alekyuk** [comment]
Hi everyone, I just discovered Positron and I feel that it could be the perfect IDE for my needs. But as many others I am using Julia more and more and therefore would love to have native support for it, allowing the plots to appear in the 'PLOTS' panel as well as 'VARIABLES' and else. I hope that this dream will come to life soon !

**Comment by donboyd5** [comment]
+1 for julia

**Comment by reikoch** [comment]
+1 for julia

**Comment by maucejo** [comment]
+1for Julia !

**Comment by prochlorothrix** [comment]
The VSC Julia extension works well in Positron (actually better than in VSC itself), although it would be nice to have a real Julia console similar to the R and Python ones.

**Comment by ntluong95** [comment]
> The VSC Julia extension works well in Positron (actually better than in VSC itself), although it would be nice to have a real Julia console similar to the R and Python ones.

Agree. But I guess Posit team is still busy this time for the launching, so we cannot expect it will come soon

**Comment by SamuelMathieu-code** [comment]
Again +1 for Julia

**Comment by davidanthoff** [comment]
> The VSC Julia extension works well in Positron (actually better than in VSC itself)

Could you elaborate a bit more how it works better in Positron than in VSC?

**Comment by juliasilge** [comment]
@davidanthoff We would be very interested in collaborating if you are interested in supporting Positron features in the VS Code extension. Although Positron's API for extensions is not fully mature yet, I can point you to:

- https://positron.posit.co/extension-development
- the Quarto extension we maintain, which supports both vanilla VS Code and Positron-specific features: https://github.com/quarto-dev/quarto/tree/main/apps/vscode

We would be very happy to review PRs or otherwise collaborate with you and the extension maintainers.

**Comment by ntluong95** [comment]
The minimum support of Julia extension to Positron I hope is it has a dedicated console, be able to run multiple sessions, plot view, variable pane, data explorer. @davidanthoff

**Comment by pbhogale** [comment]
Looking forward to native-like Julia support.

**Comment by wetlandscapes** [comment]
Having native Julia support would be a dream. It's the data science trifecta with the general capabilities of python, the statistical tools of R, and the numerical capabilities of Julia -- not to mention each supports extremely useful, but not always interoperable, paradigms related to object oriented programming, functional programming, and multiple dispatch.
I think this kind of thing could also be useful for learning a new language (looking at you julia), as it lowers the barrier to entry due to the ability to more easily pursue polyglotic projects that leverage each languages strength. My two cents.

**Comment by stefaniecg** [comment]
yes, please make Positron support Julia !
Thank you !

**Comment by hikeinsky** [comment]
Please support Julia  alongside Py and R! This will make Positron the data science killer editor. Thank you!

**Comment by GustavoGarone** [comment]
+1

**Comment by purpledawn777** [comment]
+1

**Comment by ma-galan** [comment]
This would be great! +1

**Comment by ntluong95** [comment]
This feature request has the most vote I guess. But seems they dont have a plan atm

**Comment by EvoArt** [comment]
This would be amazing! +1

**Comment by edwardlavender** [comment]
+1 for julia!

**Comment by aaronpeikert** [comment]
+1

**Comment by NilsDroste** [comment]
+1

**Comment by wesm** [comment]
☝️ Started working on this with some AI assistance. Likely will take a few weeks of work to get into a serviceable state but contributions (and more importantly bug reports) will be welcome once the initial PR lands. It will likely go behind a feature flag in the settings to turn on while in development until it is more stable and the QA team is confident that it will not conflict with the other parts of Positron if it is enabled by default in the future.

**Comment by rspreafico** [comment]
+1

**Comment by stephenfeagin** [comment]
+1

**Comment by bonStats** [comment]
+1

**Comment by nicolasfoss** [comment]
Yes, please add support for Julia!

**Comment by KSepetanc** [comment]
Keep us updated about the progress and roadmap for this. Looking forward to it very much. Last change in PR was 2 months ago so makes me wonder when is next step.

**Comment by wesm** [comment]
hi all — there isn't the appetite right now to take on the positron-julia extension I started building as a built-in extension in the main Positron repository, and so to be able to move forward someone could pick up the work that I started and develop the extension as a VSIX add-on for Positron. I don't personally have the time to continue developing it but I could assist or help facilitate if someone watching this issue wants to pick up the project.

**Comment by dolcos** [comment]
Please add support for Julia!

**Comment by ntluong95** [comment]
Hi guys, based on the work of @wesm, I created the Julia extension for Positron, which you can give it a try here https://open-vsx.org/extension/ntluong95/positron-julia

Please report any bugs or open a feature request. I am not a fulltime developer, so if anyone wants to give me a hand, you are very welcome.

**Comment by ntluong95** [comment]
For those who are still wondering what are included in Julia-Positron extension, here are some snapshots:

### Language server with diagnostics and semantic highlighting
<img width="863" height="298" alt="Image" src="https://github.com/user-attachments/assets/b26cdaae-9a1a-406e-a85b-281d8a3b67fc" />

### Access to Data Explorer and Plot pane

<img width="524" height="1214" alt="Image" src="https://github.com/user-attachments/assets/60675b0c-9006-4bd8-8d5a-eeff76992aa5" />

### Access to Help pane

<img width="512" height="1229" alt="Image" src="https://github.com/user-attachments/assets/04a40ea4-ee3f-4a8a-8353-8e1e843ec46e" />

### Access to Package pane

<img width="383" height="444" alt="Image" src="https://github.com/user-attachments/assets/45a36fb2-50b3-42dc-80ea-f6569dd7d846" />

**Comment by Amphiprion** [comment]
Saddly it didn't work to me i got the error : 

Julia 1.11.6 (x86_64) failed to start up (exit code 130)

Timed out waiting to connect to session's ZeroMQ sockets after 30 seconds
  Activating project at `c:\Users\mcout\.positron\extensions\ntluong95.positron-julia-0.0.8-universal\julia\Positron`
Precompiling IJulia...
           ✗ OpenSSL_jll
           ✗ PrecompileTools
           ✗ ZMQ
           ✗ Parsers
           ✗ JSON
           ✗ Conda
Info Given IJulia was explicitly requested, output will be shown live 
ERROR: LoadError: UndefVarError: `StaticData` not defined in `Base`
Stacktrace:
 [1] getproperty(x::Module, f::Symbol)
   @ Base .\Base.jl:42
 [2] top-level scope
   @ C:\Users\mcout\.julia\packages\PrecompileTools\gn08A\src\invalidations.jl:15
 [3] include(mod::Module, _path::String)
   @ Base .\Base.jl:562
 [4] include(x::String)
   @ PrecompileTools C:\Users\mcout\.julia\packages\PrecompileTools\gn08A\src\PrecompileTools.jl:1
 [5] top-level scope
   @ C:\Users\mcout\.julia\packages\PrecompileTools\gn08A\src\PrecompileTools.jl:16
 [6] include
   @ .\Base.jl:562 [inlined]
 [7] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::String)
   @ Base .\loading.jl:2881
 [8] top-level scope
   @ stdin:6
in expression starting at C:\Users\mcout\.julia\packages\PrecompileTools\gn08A\src\invalidations.jl:15
in expression starting at C:\Users\mcout\.julia\packages\PrecompileTools\gn08A\src\PrecompileTools.jl:1
in expression starting at stdin:6

**Comment by bemts-hhs** [comment]
@ntluong95 I'm excited to try the extension, already downloaded. I will submit feedback as I gain experience.

**Comment by juliasilge** [comment]
Thank you so much to everyone here, for your feedback and thoughts! And a special shout-out to @ntluong95 for building and publishing the [Julia extension for Positron](https://github.com/TidierOrg/positron-julia).

We've done a bit of work on our side to make an extension like this easier to maintain in the long run and more discoverable as well:

- https://github.com/TidierOrg/positron-julia/pull/41 added the ability the run tests against the Positron API and set up a GH action
- https://github.com/posit-dev/positron-website/pull/423 talks about the Julia extension as an example for such tests
- https://github.com/posit-dev/positron-website/pull/426 documents that extensions can provide additional language interpreters (uses the Julia extension and ggsql as the examples)

With these changes now, if you do the regular (non-AI) search on our docs site, you get info pointing you the Julia extension:

<img width="891" height="395" alt="Image" src="https://github.com/user-attachments/assets/72597504-d711-4121-a4c3-13d7b364f8cc" />

And if you use the AI search to find out about using Julia in Positron, you get this info now:

<img width="517" height="564" alt="Image" src="https://github.com/user-attachments/assets/355262d4-7af0-4791-84ef-9e68df992ed2" />

We are excited to continue to support community extensions like these, and welcome further reports on how to make the extension developer experience better. Thanks once again! 🙌
